### PR TITLE
Named profiles and allow default if there is only one

### PR DESCRIPTION
**Pull request by depombo**

### CODE REVIEW

File: cli/src/deploy/mod.rs [hunk 1]
```diff
@@ -170,25 +171,42 @@ fn get_url() -> &'static str {
...
+      let cred_prof_name = match profile.credentialProfile {
+        None => {
+          if cred_profs.len() != 1 {
+            let err = format!(
+              "No credential profile specified in deploy config {} when more than one \
+              credential profile exists in {}.",
+              deploy_profile_name, CREDENTIALS_FILE
+            );
+            eprintln!("{}", err);
+            error!("{}", err);
```
Comment (dfellis) [inline cli/src/deploy/mod.rs:189]: Do you need both `eprintln!` and `error!` here? I thought `error!` would also print the message?

**Review by dfellis — APPROVED** [review]
Just a small question on this.

File: cli/src/deploy/mod.rs (hunk 1 above)
Comment (depombo) [reply cli/src/deploy/mod.rs:189]: so `eprintln` displays the message to the user and `error!` logs it for us, but the semantics of this change slightly with https://github.com/alantech/anycloud/pull/41 where `error!` goes to the `clientError` endpoint

**Review by cdmistman — APPROVED** [review]

File: cli/src/deploy/mod.rs (hunk 1 above)
Comment (depombo) [reply cli/src/deploy/mod.rs:189]: We could create a utility function that calls both when the messages are the same

Comment (dfellis) [reply cli/src/deploy/mod.rs:189]: While having two backends for actual logging is weird, having one backend to a logging service and one backend to stderr makes total sense to me. I'd prefer if `error!` just splatted the message across whatever backends were connected, with `eprintln!` being a simple one that we have added to the list of backends to work with.

Comment (dfellis) [reply cli/src/deploy/mod.rs:189]: Not a blocker for this PR, though.

File: cli/src/deploy/mod.rs [hunk 2]
```diff
@@ -386,85 +404,65 @@ pub async fn info() {
...
       &app.status,
     ]);
+    for (i, profile) in app.cloudConfigs.iter().enumerate() {
+      if deploy_profiles.contains(&app.deployName) {
+        continue;
+      }
+      if i == 0 {
+        profile_data.push(vec![&app.deployName, &profile.region, &profile.vmType])
+      } else {
+        profile_data.push(vec![&"", &profile.region, &profile.vmType])
```
Comment (aguillenv) [inline cli/src/deploy/mod.rs:425]: Why empty string here?

Comment (depombo) [reply cli/src/deploy/mod.rs:425]: If a deploy profile has more than one cloud config (multi-region or multi-cloud) then this let's us show it within the same row under that deploy profile in the table

File: cli/src/deploy/mod.rs [hunk 3]
```diff
@@ -170,25 +171,42 @@ fn get_url() -> &'static str {
...
+              "No credential profile specified in deploy config {} when more than one \
+              credential profile exists in {}.",
+              deploy_profile_name, CREDENTIALS_FILE
+            );
+            eprintln!("{}", err);
+            error!("{}", err);
+            client_error("INVALID_DEFAULT_CREDENTIAL_ALIAS").await;
+            std::process::exit(1);
+          }
+          cred_profs.keys().next().unwrap().to_string()
```
[the remark this answers is not in the document]
Comment (depombo) [reply cli/src/deploy/mod.rs:193]: The conditional right before fails if the `len !=1` and it is trying to fetch the default so here we safely take the first one which is the only one

[the remark this answers is not in the document]
Comment (dfellis) [reply cli/src/deploy/mod.rs:193]: First and only one. It errors if there's more than one. Rust's hashmap is like Go's in that the hashmap iteration order is not guaranteed to be insertion order.

**Pull request merged**